torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -703,10 +644,23 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
   VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is sending request #"
           << messageId << " to " << clientPipe.pipe_->getRemoteName();
 
+  auto devices = getDevicesForTensors(
+      clientPipe.pipe_->getRemoteName(), requestMessage);
+
+  // Increase clientActiveCalls_ and pass it to pipeWrite and pipeRead
+  // callbacks. It will decrease clientActiveCalls_ accordingly on destruction.
+  auto clientCallCounter =
+      std::make_shared<CallCounter>(*this, clientActiveCalls_);
```

Comment:
Could we move it to the same exact spot where we did the increase before? I am a bit worried about a race condition with the timeout thread, which could fire and decrease the count before we reach this point (see also my comment further down about the timeout thread currently not decreasing the count).

Also, it is a shame that we need to wrap this in a shared_ptr and thus perform an extra heap allocation just for this small thing. Unfortunately I suspect it's necessary because even if we made the CallCounter a full RAII object, it would be movable-only, and we cannot capture movable-only objects in std::functions. (On the TensorPipe side we are thinking of rolling our own non-copyable function wrapper which we could then use here too...)